metricflow/test/integration/test_cases/itest_metrics.yaml

```diff
@@ -863,3 +863,142 @@ integration_test:
       ) b
       ON a.metric_time__day = b.metric_time__day
     ) c
+---
+integration_test:
+  name: derived_metric_offset_window_and_granularity
+  description: Tests a derived metric offset query with window and granularity
+  model: SIMPLE_MODEL
+  metrics: ["bookings_5_day_lag"]
+  group_bys: ["metric_time__month"]
+  check_query: |
+    SELECT
+      {{ render_date_trunc("a.ds", TimeGranularity.MONTH) }} AS metric_time__month
+      , SUM(b.bookings_5_day_lag) AS bookings_5_day_lag
+    FROM {{ mf_time_spine_source }} a
+    INNER JOIN (
+      SELECT
+        ds AS metric_time__day
+        , 1 AS bookings_5_day_lag
+      FROM {{ source_schema }}.fct_bookings
+    ) b
+    ON {{ render_date_sub("a", "ds", 5, TimeGranularity.DAY) }} = b.metric_time__day
+    GROUP BY metric_time__month
+---
+integration_test:
+  name: derived_metric_with_offset_to_grain_and_granularity
+  description: Tests a derived metric query with an offset_to_grain and granularity
+  model: SIMPLE_MODEL
+  metrics: ["bookings_growth_since_start_of_month"]
+  group_bys: ["metric_time__week"]
+  check_query: |
+    SELECT
+      COALESCE(a.metric_time__week, b.metric_time__week)  AS metric_time__week
+      , bookings - bookings_at_start_of_month AS bookings_growth_since_start_of_month
+    FROM (
+      SELECT
+        {{ render_date_trunc("ds", TimeGranularity.WEEK) }} AS metric_time__week
+        , SUM(1) AS bookings
+      FROM {{ source_schema }}.fct_bookings
+      GROUP BY metric_time__week
+    ) a
+    INNER JOIN (
+      SELECT
+        {{ render_date_trunc("c.ds", TimeGranularity.WEEK) }} AS metric_time__week
+        , SUM(d.bookings_at_start_of_month) AS bookings_at_start_of_month
+      FROM {{ mf_time_spine_source }} c
+      INNER JOIN (
+        SELECT
+          ds AS metric_time__day
+          , 1 AS bookings_at_start_of_month
+        FROM {{ source_schema }}.fct_bookings
+      ) d
+      ON {{ render_date_trunc("c.ds", TimeGranularity.MONTH) }} = d.metric_time__day
+      GROUP BY metric_time__week
+    ) b
+    ON a.metric_time__week = b.metric_time__week
```

Comment:
I think we need something like this, or else to do aggregation first and then join as we were doing before, but somehow to propagate the granularity spec down for the offset scenario:

```suggestion
    SELECT
      COALESCE(a.metric_time__week, b.metric_time__week)  AS metric_time__week
      , bookings - bookings_at_start_of_month AS bookings_growth_since_start_of_month
    FROM (
      SELECT
        {{ render_date_trunc("ds", TimeGranularity.WEEK) }} AS metric_time__week
        , SUM(1) AS bookings
      FROM {{ source_schema }}.fct_bookings
      GROUP BY metric_time__week
    ) a
    INNER JOIN (
      SELECT
        {{ render_date_trunc("c.ds", TimeGranularity.WEEK) }} AS metric_time__week
        , SUM(d.bookings_at_start_of_month) AS bookings_at_start_of_month
      FROM {{ mf_time_spine_source }} c
      INNER JOIN (
        SELECT
          ds AS metric_time__day
          , 1 AS bookings_at_start_of_month
        FROM {{ source_schema }}.fct_bookings
      ) d
      ON {{ render_date_trunc("c.ds", TimeGranularity.MONTH) }} = d.metric_time__day
      -- filter time spine join output to the dates we need to carry forward
      WHERE {{ render_date_trunc("c.ds", TimeGranularity.WEEK) }} = c.ds
      GROUP BY metric_time__week
    ) b
    ON a.metric_time__week = b.metric_time__week
```

Reply:
Had to do a bunch of testing on real SQL results to wrap my head around this but can confirm that that filter works! Question though - does it make a difference if we do the filter in the `WHERE` clause vs the `JOIN` clause? They appear to behave the same way so I'm wondering if one is more efficient.

Reply:
Also, this complicates things if you query with multiple granularities. So I need to do some more SQL testing to figure out what syntax we want in that case.

Reply:
So, we could do an `OR` statement for multiple granularities, like:
`WHERE DATE_TRUNC(MONTH, ds) = ds OR DATE_TRUNC(WEEK, ds) = ds`
 That results in a result set like this for a metric with offset to grain:
![image](https://github.com/dbt-labs/metricflow/assets/26731294/53c6fc64-87ec-4b31-8397-30f1637e5bda)
I haven't thought of a better answer than this.

Reply:
I went ahead and implemented it in the `WHERE` clause and with the `OR` statement for multiple granularities. We can go through the generated SQL and make sure that all makes the most sense!

Reply:
Doing it in the JOIN clause should be more efficient, but I don't know if all engines support that construction. Let's stick with the WHERE for now and we can look into optimizing later.